web/src/reducers/budget.js

```diff
@@ -91,320 +90,331 @@ const initialState = years => ({
+        activityTotals.data.combined.total += cost;
       });
     });
 
-    // Same with contractor expenses, but this one is straightforward.
-    const contractorExpenses = totaller.collapse('contractorResources').raw();
-
-    // This is the percent of the federal share per fiscal quarter,
-    // by expense type.  This is entered in the
-    // ActivityQuarterlyBudgetSummary table
-    const ffpPercents = activity.quarterlyFFP;
-
-    // The grand total of the federal share of this activity's
-    // state and contractor expenses, across all quarters
-    // and fiscal years
-    const total = { state: 0, contractors: 0, combined: 0 };
-
-    newState.activities[activity.key].quarterlyFFP = Object.keys(
-      stateExpenses
-    ).reduce(
-      (ffyCumulate, year) => ({
-        ...ffyCumulate,
-        [year]: [1, 2, 3, 4].reduce(
-          (quarterCumulatve, quarter) => {
-            // The federal percent for this FFY, per quarter
-            const ffyQuarterPct = ffpPercents[year][quarter];
-
-            // These are the federal shares of these expenses
-            // for the fiscal quarter.  The budget reducer
-            // stores percentages as 0-1, so we need to
-            // convert those here.
-
-            const statePct = ffyQuarterPct.state / 100;
-            const state = stateExpenses[year].federal * statePct;
-
-            const contractorsPct = ffyQuarterPct.contractors / 100;
-            const contractors =
-              contractorExpenses[year].federal * contractorsPct;
-
-            const combined = state + contractors;
-
-            // Capture this in the grand total...
-            total.state += state;
-            total.contractors += contractors;
-            total.combined += combined;
-
-            // ...and in the FFY subtotal
-            const { subtotal } = quarterCumulatve;
-            subtotal.state.dollars += state;
-            subtotal.state.percent += statePct;
-            subtotal.contractors.dollars += contractors;
-            subtotal.contractors.percent += contractorsPct;
-            subtotal.combined.dollars += combined;
-
-            return {
-              ...quarterCumulatve,
-              [quarter]: {
-                state: { dollars: state, percent: statePct },
-                contractors: { dollars: contractors, percent: contractorsPct },
-                combined: { dollars: combined, percent: 0 }
-              },
-              subtotal
-            };
-          },
-          {
-            subtotal: {
-              state: { dollars: 0, percent: 0 },
-              contractors: { dollars: 0, percent: 0 },
-              combined: { dollars: 0, percent: 0 }
-            }
-          }
-        )
-      }),
-      {}
-    );
-    newState.activities[activity.key].quarterlyFFP.total = total;
-  });
+    activity.statePersonnel.forEach(person => {
+      Object.entries(person.years).forEach(([year, { amt, perc }]) => {
+        const cost = amt * perc / 100;
```

Comment:
can we assume this calculation will work? could `amt` or `perc` be blank or null?